### DISCUSSION TITLE
What does "detecting compiler hash" actually do?

### DISCUSSION THREAD

**Discussion by lanyizi** [microsoft/vcpkg#26999] [Q&A]
Hi everyone!
I have a CMake project which is using vcpkg with manifest mode (i.e. with `vcpkg.json`).

Each time I need to re run CMake (for example because I edited `CMakeLists.txt` to add some new source files), `vcpkg install` is automatically invoked. Since my libraries are all already installed, it won't rebuild my dependencies, so, theoretically, it shouldn't take too much time.

But it actually takes a lot! Each time I re run CMake, `vcpkg install` takes like 10 seconds to complete. I noticed that most of time it's "detecting compiler hash", what does that actually do? Is there any way to speed it up?

The following CMake output shows vcpkg is taking 9.109 s to install:
```
1> 命令行: "C:\WINDOWS\system32\cmd.exe" /c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Visual Studio 17 2022" -A Win32  -DCMAKE_TOOLCHAIN_FILE:FILEPATH="C:\Users\lanyi\source\repos\vcpkg/scripts/buildsystems/vcpkg.cmake" -DVCPKG_TARGET_TRIPLET:STRING="x86-windows-static" -DCMAKE_BUILD_TYPE:STRING="Release" -DCMAKE_INTERPROCEDURAL_OPTIMIZATION:STRING="TRUE" -DCMAKE_INSTALL_PREFIX:PATH="D:/lanyi/Source/Repos/NativeDll/out/install/x86-release"  "D:\lanyi\Source\Repos\NativeDll" 2>&1"
1> 工作目录: D:/lanyi/Source/Repos/NativeDll/out/build/x86-release
1> [CMake] -- Running vcpkg install
1> [CMake] Detecting compiler hash for triplet x86-windows-static...
1> [CMake] All requested packages are currently installed.
1> [CMake] Restored 0 packages from C:\Users\lanyi\AppData\Local\vcpkg\archives in 1.238 ms. Use --debug to see more details.
1> [CMake] 
1> [CMake] Total elapsed time: 9.109 s
1> [CMake] 
1> [CMake] The package boost is compatible with built-in CMake targets:
1> [CMake] 
1> [CMake]     find_package(Boost REQUIRED [COMPONENTS <libs>...])
1> [CMake]     target_link_libraries(main PRIVATE Boost::boost Boost::<lib1> Boost::<lib2> ...)
1> [CMake] 
1> [CMake] cpp-httplib is header-only and can be used from CMake via:
1> [CMake] 
1> [CMake]     find_path(CPP_HTTPLIB_INCLUDE_DIRS "httplib.h")
1> [CMake]     target_include_directories(main PRIVATE ${CPP_HTTPLIB_INCLUDE_DIRS})
1> [CMake] 
1> [CMake] The package fmt provides CMake targets:
1> [CMake] 
1> [CMake]     find_package(fmt CONFIG REQUIRED)
1> [CMake]     target_link_libraries(main PRIVATE fmt::fmt)
1> [CMake] 
1> [CMake]     # Or use the header-only version
1> [CMake]     find_package(fmt CONFIG REQUIRED)
1> [CMake]     target_link_libraries(main PRIVATE fmt::fmt-header-only)
1> [CMake] 
1> [CMake] The package nlohmann-json provides CMake targets:
1> [CMake] 
1> [CMake]     find_package(nlohmann_json CONFIG REQUIRED)
1> [CMake]     target_link_libraries(main PRIVATE nlohmann_json::nlohmann_json)
1> [CMake] 
1> [CMake] The package nlohmann-json can be configured to not provide implicit conversions via a custom triplet file:
1> [CMake] 
1> [CMake]     set(nlohmann-json_IMPLICIT_CONVERSIONS OFF)
1> [CMake] 
1> [CMake] For more information, see the docs here:
1> [CMake]     
1> [CMake]     https://json.nlohmann.me/features/macros/#json_use_implicit_conversions
1> [CMake] 
1> [CMake] wil provides CMake targets:
1> [CMake]     # this is heuristically generated, and may not be correct
1> [CMake]     find_package(wil CONFIG REQUIRED)
1> [CMake]     target_link_libraries(main PRIVATE WIL::WIL)
1> [CMake] 
1> [CMake] -- Running vcpkg install - done
1> [CMake] -- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19044.
1> [CMake] -- Configuring done
1> [CMake] -- Generating done
```

**Comment by JackBoosY** [comment]
ABI info which contains compiler version, SDK version etc.

**Comment by bscottm** [comment]
In a nutshell, fingerprinting the compiler allows `vcpkg` to separate different API and compiler configurations. For example, if you have VS 2022, VS 2019 and GCC installed, you want three different dependency packages that correspond to each of your three compilers. Otherwise, you potentially run into interesting compile-time and link-time problems when GCC doesn't produce the same ABI code that VS 2022 expects to link with. These are hard problems to diagnose.

**Comment by lord-ne** [comment]
Is there a way to make this faster? It takes 20-30 seconds on my machine (WSL), but I saw it taking less than a second on someone else's computer (Mac), while the rest of the compilation was about the same speed

**Reply by Osyotr** [reply]
You can try to disable windows interop in wsl settings. I remember it was causing some rediculous slowdowns on my machine due to PATH injection.
Also, https://github.com/microsoft/vcpkg-tool/pull/362

**Reply by sketch34** [reply]
Just to clarify for others, for me it was setting `appendWindowsPath=false` that fixed the slow "Detecting compiler hash" step.

My complete `wsl.conf`:

```
[boot]
systemd=true

[interop]
enabled=false
appendWindowsPath=false
```

**Comment by silverqx** [comment]
I have exactly the same problem on GitHub self-hosted runner (Windows), it lasts 100 seconds to finish `Detecting compiler hash for triplet x64-windows...`, I tried to disable WSL interop like

```
[interop]
enabled=false
appendWindowsPath=false
```

I did `Restart-Service LxssManager` but it didn't help. Do you have any idea what else can cause this?
